sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -537,9 +536,55 @@ class Dataset[T] private[sql](
    */
   @Experimental
   @InterfaceStability.Evolving
-  def checkpoint(eager: Boolean): Dataset[T] = {
+  def checkpoint(eager: Boolean = true): Dataset[T] = _checkpoint(eager = eager)
+
+  /**
+   * Eagerly locally checkpoints a Dataset and return the new Dataset. Checkpointing can be
+   * used to truncate the logical plan of this Dataset, which is especially useful in iterative
+   * algorithms where the plan may grow exponentially. Local checkpoints are written to executor
+   * storage and despite potentially faster they are unreliable and may compromise job completion.
+   *
+   * @group basic
+   * @since 2.3.0
+   */
+  @Experimental
+  @InterfaceStability.Evolving
+  def localCheckpoint(): Dataset[T] = _checkpoint(eager = true, local = true)
+
+  /**
+   * Locally checkpoints a Dataset and return the new Dataset. Checkpointing can be used to truncate
+   * the logical plan of this Dataset, which is especially useful in iterative algorithms where the
+   * plan may grow exponentially. Local checkpoints are written to executor storage and despite
+   * potentially faster they are unreliable and may compromise job completion.
+   *
+   * @group basic
+   * @since 2.3.0
+   */
+  @Experimental
+  @InterfaceStability.Evolving
+  def localCheckpoint(eager: Boolean = true): Dataset[T] = _checkpoint(eager = eager, local = true)
+
+  /**
+   * Returns a checkpointed version of this Dataset. Checkpointing can be used to truncate the
+   * logical plan of this Dataset, which is especially useful in iterative algorithms where the
+   * plan may grow exponentially.
+   * By default reliable checkpoints are created and saved to files inside the checkpoint
+   * directory set with `SparkContext#setCheckpointDir`. If local is set to true a local checkpoint
+   * is performed instead. Local checkpoints are written to executor storage and despite
+   * potentially faster they are unreliable and may compromise job completion.
+   *
+   * @group basic
+   * @since 2.3.0
+   */
+  @Experimental
+  @InterfaceStability.Evolving
+  private[sql] def _checkpoint(eager: Boolean, local: Boolean = false): Dataset[T] = {
```

Comment:
I changed according to the first review. I also agree to be private so that "there's only one way of doing it" when using the API. But if you have a strong feeling I can surely change

Reply:
I guess we have 2 options here:

- expose `def checkpoint(eager: Boolean, local: Boolean): Dataset[T]` as public, which can be used similar to `localCheckpoint`.
- make `def _checkpoint(eager: Boolean, local: Boolean = false): Dataset[T]` private to be used only from the public APIs.

and I'm afraid the current one is not good anyway.

I'd prefer the second option but I don't have a strong feeling.
cc @felixcheung @HyukjinKwon

Reply:
Sounds good to me to remove default param values, keeping private. Let me know

Reply:
`private checkpoint` makes sense to me too. i think we don't need the underscore though as it's going to have `private` ahead and strictly it might be redundant to have the leading underscore to note it's private is.

Reply:
Alright, so I remove both default values, otherwise we end up with a colliding signature.